### ISSUE TITLE
Find way to properly share code between the main repo and this repo

### ISSUE THREAD

**Issue by davidebbo** [OneZoom/tree-build#6]
The one case I can see today is where OTT_popularity_mapping.py imports `wikiflags`:

```py
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), os.path.pardir, os.path.pardir, os.path.pardir, "models")))
from _OZglobals import wikiflags
```

This becomes harder with the tree building code in a separate repository.

If it's just `wikiflags`, we can just duplicate it, but if there are more significant things that need sharing, we may need to come up with a better solution. e.g. it could be moved to a separate OZGlobals module that would be imported by both repos.

**Comment by hyanwong** [comment]
Ah, good point.

I think it's only that specific set of flags that are used in the popularity code. They need to be in sync with the definitions used in the DB though, so making a copy means that we risk them becoming out-of-sync. Since we have some unit tests in the main OneZoom repo, we could, I suppose, check in one of those that the files are in sync somehow.

**Comment by davidebbo** [comment]
@hyanwong yes, we could do something like that. Another possible approach could be to not commit the file to this repo, but instead download it to this repo using some setup script, and gitignore it. Either way, it feels a bit dirty.

**Comment by hyanwong** [comment]
Just returning to this issue. Other options would be:

1.  define the mappings in a config file created by the tree-build repo, which is read into the alongside the CSV files, and somehow read this config file into the web2py instance instead of hard-coding it.
2. add the mappings to a new field in the database, which is populated in e.g. step 9 of https://github.com/OneZoom/tree-build/blob/main/oz_tree_build/README.markdown

I don't know the efficiency issues associated with either of these versus the hard-coded `cache.ram()` globals that we are currently using.

**Comment by hyanwong** [comment]
Another (slightly less complicated) case is where there are hard-coded OTT ids on the main OneZoom website (for example, in https://github.com/OneZoom/OZtree/blob/main/views/default/footer_sponsor_items.load). After building a tree, we should check that these OTT numbers are all still present on the tree.

The mos sensible way to do this, IMO, would be for the OZtree repository to have a centralised dictionary that maps each name (e.g. "seahorses") to an OTT. Then we can provide some command-line instructions to load this dictionary and check the numbers are still in the tree.

An obvious place to put a centralised dictionary of this sort is in _OZglobals, as we need that for the wikiflags data anyway.

**Comment by davidebbo** [comment]
Yes, if had this in a shared globals accessible by both repos, it would be simple to have a command line tool that checks for their presence in a given tree.

Looking at OZtree, one challenge is that these OTT ids currently sit in this hard coded footer_sponsor_items.load file. Instead, this part of the file would need to be templated/generated based on the centralised dictionary, so they only exist in one place.

**Comment by hyanwong** [comment]
Yes, but that `footer_sponsor_items.load` file has access to the globals defined in `_OZglobals`, so it would be easy to do this without requiring a DB access command in the footer, which was the reason to hard-code them

**Comment by davidebbo** [comment]
Agreed.

If the goal is to share a .py file between the two repos, another possibility we haven't discussed is to use a git submodule. Essentially, we'd have any shared files in a new `oz-shared` repo, and both `tree-build` and `OZtree` would include it in some subfolder as a 'git submodule'. I've used those before. The commands are a little tricky to understand, but they do work.

One benefit is that it's a more source control friendly solution, in the sense that different branches/commits of the two main repos can be mapped to different commits on the shared submodules.

**Comment by hyanwong** [comment]
Just thinking about this again. As you say, submodules have the advantage that you can use different ones for different tree versions. For example, the species in the footer are very specific to the current OneZoom tree: they might not exist at all in other versions.

A final possibility is to put the OTT values into the appconfig.ini file, which can be altered for different setups and I think is cached when "is_testing = False".

**Comment by davidebbo** [comment]
I feel this issue is now tracking two somewhat different things:
1. Sharing **code** that needs to be used and in sync between both repos. `wikiflags` is an example of this. It does not change based on tree. I think a submodule is a reasonable solution for this.
2. Sharing tree specific **data**, like what's in `footer_sponsor_items.load`. Here, storing in config or DB makes sense, as it is effectively a form of configuration.

> Another possibility, but which wouldn't allow easy sharing between repos, is to check via an API on the OneZoom website

I'm trying to understand the end to end flow that you have in mind. Would logic in the `tree-build` repo call this API? Does that create a bit of a chicken and egg situation in the sense that you can't run a OneZoom site until you have built a tree using `tree-build`? I'm probably misunderstanding.

**Comment by davidebbo** [comment]
Completely agree about submodules. They work, but they commands to deal with them are quite counterintuitive. If it's just for `wikiflags`, maybe it's simpler to just accept the small duplication and not do anything else. If we find we need to share more code, we can bring submodules back into the conversation.

> There are various things on the OneZoom website that assume a set a taxa exist on the tree (e.g. sponsorships for one!). It would be sensible to have a way to check this after building a tree, and, having thought about it a bit, I suppose an API query to the main site would be the most sensible. This would also, I think, mean that the best solution to the foster would be to place the footer items in the DB and simply cache the DB results so that the footer only has to access the DB e.g. once per day or week.

That sounds reasonable. As long as there is some easy way to get a list of OTTs that must be in the tree, checking for them is easy. This logic could actually be in either repo, i.e.
1. If in tree-build, you'd run it after building a tree
2. If in OZtree, you'd run it before importing the new `ordered_leaves`/`ordered_nodes` SQL.
Either way, it can easily run against the ordered_nodes_[version].csv.

It would also make sense to decouple getting the list from verifying the presence into different scripts (that can be piped). In fact, the first part could just be a curl command.

I would suggest opening a separate OZtree issue to track the work to moves things into `tree_startpoints`, and exposes a new API. That actually seems like the harder work. Once we have that, the verification script is quite simple.

**Comment by lentinj** [comment]
> I would suggest opening a separate OZtree issue to track the work to moves things into tree_startpoints

My preference would be tree-build populating the ``tree_startpoints`` table along with the rest of it's DB dump, that added values that make sense for the tree that tree-build is building. However, the getting-into-tree_startpoints and rely-on-caching and git-submodules-are-terrible points above I can all get behind.

However, what happens to things like [AT.html](https://github.com/OneZoom/OZtree/blob/main/views/treeviewer/AT.html) and it's custom locations? I think living in ``AT.html`` with the rest of the customisations that make the page makes sense, but obviously that's not an API call you can scan. Worse, what if an embed has a bunch of configured startpoints in it's URL? You're definitely not scanning those.

I think there needs to be a means to specify that AT needs a named tree, and doesn't make sense without. This also applies to tours I think, You could have tours that make sense for the "life" tree but not the "extinct" tree, load them all and show whichever make sense.

**Comment by davidebbo** [comment]
> My preference would be tree-build populating the `tree_startpoints` table along with the rest of it's DB dump, that added values that make sense for the tree that tree-build is building.

Yes, I think that does make sense. The full flow would be something like:
- tree-build has a data file that captures the start point, and these are tree specific, just like the newick file
- tree-build turns it into a SQL file, as it does with leaves/nodes
- it all gets imported, probably truncating the existing startpoint table first. There should never be a need to manually update the startpoint table, just like we never directly mess with the leaf/node tables.

> This also applies to tours I think, You could have tours that make sense for the "life" tree but not the "extinct" tree, load them all and show whichever make sense

I think this goes right back into the discussion of how we'll want to handle multiple trees. If each tree has its own DB (along with separate web2py apps for each), then there is no need to 'load them all and show whichever make sense', because the `tree_startpoints` instance will always be specific to the current tree. To me, this feels simpler, as it keeps all the app logic single-tree, and just relies on parallel universes to handle more trees (and yes, this breaks down a bit for root site pages that need to present the multiple trees).

**Comment by lentinj** [comment]
Yeah, keeping ``tree_startpoints`` single-tree would definitely be easier. I was referring to the ``tours`` table here, but OTOH maybe that's also single-tree, and the structures in the tours repository have an assigned tree. If the tree they expect doesn't match, the tour doesn't load.

> If each tree has its own DB (along with separate web2py apps for each)

I don't think it's going to work long term, there'll eventually be a requirement to share sponsorships between trees for example, but It's a good place to start and can worry about supporting multiple trees later.
